File: forgerock-auth/src/test/java/org/forgerock/android/auth/callback/DeviceProfileCollectorCallbackTest.java
/*
 * Copyright (c) 2019 - 2022 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth.callback;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

import android.content.Context;

import androidx.test.core.app.ApplicationProvider;

import org.forgerock.android.auth.Config;
import org.forgerock.android.auth.FRListenerFuture;
import org.forgerock.android.auth.KeyStoreManager;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import java.security.PublicKey;

@RunWith(RobolectricTestRunner.class)
public class DeviceProfileCollectorCallbackTest {

    private Context context = ApplicationProvider.getApplicationContext();

    @Mock
    public KeyStoreManager keyStoreManager;

    @Mock
    public PublicKey publicKey;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        byte[] encoded = "public key".getBytes();
        when(publicKey.getEncoded()).thenReturn(encoded);
        when(keyStoreManager.getIdentifierKey(any())).thenReturn(publicKey);
        Config.getInstance().init(context, null);
        Config.getInstance().setKeyStoreManager(keyStoreManager);
    }

    @Test
    public void testMetadata() throws Exception {

        JSONObject raw = new JSONObject("{\n" +
                "            \"type\": \"DeviceProfileCallback\",\n" +
                "            \"output\": [\n" +
                "                {\n" +
                "                    \"name\": \"metadata\",\n" +
                "                    \"value\": true\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"location\",\n" +
                "                    \"value\": false\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"message\",\n" +
                "                    \"value\": \"\"\n" +
                "                }\n" +
                "            ],\n" +
                "            \"input\": [\n" +
                "                {\n" +
                "                    \"name\": \"IDToken1\",\n" +
                "                    \"value\": \"\"\n" +
                "                }\n" +
                "            ]\n" +
                "        }");
        DeviceProfileCallback callback = new DeviceProfileCallback(raw, 0);
        FRListenerFuture<Void> result = new FRListenerFuture<>();
        callback.execute(context, result);
        result.get();

        String content = ((JSONObject) callback.getContentAsJson().getJSONArray("input").get(0)).getString("value");

        JSONObject contentAsJson = new JSONObject(content);
        assertThat(contentAsJson.get("identifier")).isNotNull();
        assertThat(contentAsJson.get("metadata")).isNotNull();
    }

    @Test
    public void testNoAttributesToCollect() throws Exception {

        JSONObject raw = new JSONObject("{\n" +
                "            \"type\": \"DeviceProfileCallback\",\n" +
                "            \"output\": [\n" +
                "                {\n" +
                "                    \"name\": \"metadata\",\n" +
                "                    \"value\": false\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"location\",\n" +
                "                    \"value\": false\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"message\",\n" +
                "                    \"value\": \"\"\n" +
                "                }\n" +
                "            ],\n" +
                "            \"input\": [\n" +
                "                {\n" +
                "                    \"name\": \"IDToken1\",\n" +
                "                    \"value\": \"\"\n" +
                "                }\n" +
                "            ]\n" +
                "        }");
        DeviceProfileCallback callback = new DeviceProfileCallback(raw, 0);
        FRListenerFuture<Void> result = new FRListenerFuture<>();
        callback.execute(context, result);
        result.get();

        String content = ((JSONObject) callback.getContentAsJson().getJSONArray("input").get(0)).getString("value");

        JSONObject contentAsJson = new JSONObject(content);
        assertThat(contentAsJson.get("identifier")).isNotNull();
        assertThat(contentAsJson.opt("metadata")).isNull();
        assertThat(contentAsJson.opt("location")).isNull();

    }


    @Test
    public void testAttributesToCollect() throws Exception {

        JSONObject raw = new JSONObject("{\n" +
                "            \"type\": \"DeviceProfileCallback\",\n" +
                "            \"output\": [\n" +
                "                {\n" +
                "                    \"name\": \"metadata\",\n" +
                "                    \"value\": true\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"location\",\n" +
                "                    \"value\": true\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"message\",\n" +
                "                    \"value\": \"Test Message\"\n" +
                "                }\n" +
                "            ],\n" +
                "            \"input\": [\n" +
                "                {\n" +
                "                    \"name\": \"IDToken1\",\n" +
                "                    \"value\": \"\"\n" +
                "                }\n" +
                "            ]\n" +
                "        }");

        DeviceProfileCallback callback = new DeviceProfileCallback(raw, 0);
        assertThat(callback.isMetadata()).isTrue();
        assertThat(callback.isLocation()).isTrue();
        assertThat(callback.getMessage()).isEqualTo("Test Message");

    }


}

File: forgerock-auth/src/test/java/org/forgerock/android/auth/callback/DeviceProfileCollectorCallbackTest.kt
/*
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 *  This software may be modified and distributed under the terms
 *  of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth.callback

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.assertj.core.api.Assertions
import org.forgerock.android.auth.Config
import org.forgerock.android.auth.FRListenerFuture
import org.forgerock.android.auth.KeyStoreManager
import org.json.JSONObject
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.MockitoAnnotations
import java.security.PublicKey

@RunWith(AndroidJUnit4::class)
class DeviceProfileCollectorCallbackTest {
    private val context = ApplicationProvider.getApplicationContext<Context>()

    @Mock
    var keyStoreManager: KeyStoreManager? = null

    @Mock
    var publicKey: PublicKey? = null
    @Before
    @Throws(Exception::class)
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        val encoded = "public key".toByteArray()
        Mockito.`when`(publicKey!!.encoded).thenReturn(encoded)
        Mockito.`when`(keyStoreManager!!.getIdentifierKey(ArgumentMatchers.any()))
            .thenReturn(publicKey)
        Config.getInstance().init(context, null)
        Config.getInstance().keyStoreManager = keyStoreManager
    }

    @Test
    @Throws(Exception::class)
    fun testMetadata() {
        val raw = JSONObject("""{
            "type": "DeviceProfileCallback",
            "output": [
                {
                    "name": "metadata",
                    "value": true
                },
                {
                    "name": "location",
                    "value": false
                },
                {
                    "name": "message",
                    "value": ""
                }
            ],
            "input": [
                {
                    "name": "IDToken1",
                    "value": ""
                }
            ]
        }""")
        val callback = DeviceProfileCallback(raw, 0)
        val result = FRListenerFuture<Void>()
        callback.execute(context, result)
        result.get()
        val content =
            (callback.contentAsJson.getJSONArray("input")[0] as JSONObject).getString("value")
        val contentAsJson = JSONObject(content)
        Assertions.assertThat(contentAsJson["identifier"]).isNotNull
        Assertions.assertThat(contentAsJson["metadata"]).isNotNull
    }

    @Test
    @Throws(Exception::class)
    fun testNoAttributesToCollect() {
        val raw = JSONObject("""{
            "type": "DeviceProfileCallback",
            "output": [
                {
                    "name": "metadata",
                    "value": false
                },
                {
                    "name": "location",
                    "value": false
                },
                {
                    "name": "message",
                    "value": ""
                }
            ],
            "input": [
                {
                    "name": "IDToken1",
                    "value": ""
                }
            ]
        }""")
        val callback = DeviceProfileCallback(raw, 0)
        val result = FRListenerFuture<Void>()
        callback.execute(context, result)
        result.get()
        val content =
            (callback.contentAsJson.getJSONArray("input")[0] as JSONObject).getString("value")
        val contentAsJson = JSONObject(content)
        Assertions.assertThat(contentAsJson["identifier"]).isNotNull
        Assertions.assertThat(contentAsJson.opt("metadata")).isNull()
        Assertions.assertThat(contentAsJson.opt("location")).isNull()
    }

    @Test
    @Throws(Exception::class)
    fun testAttributesToCollect() {
        val raw = JSONObject("""{
            "type": "DeviceProfileCallback",
            "output": [
                {
                    "name": "metadata",
                    "value": true
                },
                {
                    "name": "location",
                    "value": true
                },
                {
                    "name": "message",
                    "value": "Test Message"
                }
            ],
            "input": [
                {
                    "name": "IDToken1",
                    "value": ""
                }
            ]
        }""")
        val callback = DeviceProfileCallback(raw, 0)
        Assertions.assertThat(callback.isMetadata).isTrue
        Assertions.assertThat(callback.isLocation).isTrue
        Assertions.assertThat(callback.message).isEqualTo("Test Message")
    }
}